ledger/src/blocktree.rs

```diff
@@ -5133,4 +5181,49 @@ pub mod tests {
         let num_coding_in_index = index.coding().num_coding();
         assert_eq!(num_coding_in_index, num_coding);
     }
+
+    #[test]
+    fn test_duplicate_slot() {
+        let slot = 0;
+        let entries1 = make_slot_entries_with_transactions(1);
+        let entries2 = make_slot_entries_with_transactions(1);
+        let leader_keypair = Arc::new(Keypair::new());
+        let shredder = Shredder::new(slot, 0, 1.0, leader_keypair.clone(), 0, 0)
+            .expect("Failed in creating shredder");
+        let (shreds, _, _) = shredder.entries_to_shreds(&entries1, true, 0);
+        let (duplicate_shreds, _, _) = shredder.entries_to_shreds(&entries2, true, 0);
+        let shred = shreds[0].clone();
+        let duplicate_shred = duplicate_shreds[0].clone();
+        let non_duplicate_shred = shred.clone();
+
+        let blocktree_path = get_tmp_ledger_path!();
+        {
+            let blocktree = Blocktree::open(&blocktree_path).unwrap();
+            blocktree
+                .insert_shreds(vec![shred.clone()], None, false)
+                .unwrap();
+
+            // No duplicate shreds exist yet
+            assert!(!blocktree.is_duplicate_slot(slot));
+
+            // Check if shreds are duplicated
+            assert!(blocktree.is_shred_duplicate(slot, 0, &duplicate_shred.payload));
```

Comment:
Why wouldn't blocktree itself check for this duplicate when it's storing the shred? Kind of weird that blocktree owns this comparison function which the user calls and then manually does the duplication insert.

Reply:
@sakridge I didn't want the insert logic to be blocked by a read from disk of the potentially duplicate block, and then a verification that the block was actually duplicate. I was thinking as soon as the insertion logic detects a duplicate (shred slot, shred index), it calls a handler which passes that shred to a different thread which will actually do the verification

Reply:
ok; sure